### PR TITLE
Mysql integration in Canyon

### PR DESCRIPTION
**Pull request by 0nSystem**
Hi guys, I am opening this PR as I want to do mysql integration. I open the PR in draft mode to be able to consult details and to be able to make a periodical follow-up of my progress.

Greetings ZeroDayCode.

### CODE REVIEW

File: docker/mysql/create_tables.sql [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+	last_name			TEXT NOT NULL,
+	summoner_name		TEXT NOT NULL,
+	image_url			TEXT,
+	role				TEXT NOT NULL
+);
+
+CREATE TABLE public.team (
+    id					INT AUTO_INCREMENT PRIMARY KEY,
+	ext_id				BIGINT NOT NULL,
+    slug				TEXT NOT NULL,
```
Comment (Pyzyryab) [inline docker/mysql/create_tables.sql:36]: Remember to left the code in a well formmated state

**Comment by Pyzyryab** [comment]
> Hi guys, I am opening this PR as I want to do mysql integration. I open the PR in draft mode to be able to consult details and to be able to make a periodical follow-up of my progress.
> 
> Greetings ZeroDayCode.

I am happy to see you contributing for the very first time to the Canyon's codebase. I am looking forward to see how the implementation of `MySql` fits in our project :)

**Comment by 0nSystem** [comment]
Good morning,
I am starting to create test runs of select and update macros, without starting in querybuilder.
I receive syntax errors on certain occasions, an example is in the names of the columns due to a quotation marks, in case of sql server does not affect, the upper or lower case, in case mysq_async does not allow me to keep them, and in case of postgresql if it can be the case and can be problematic, so I comment this problem to obtain a solution as effective as possible and not a horrible patch without having analyzed together.

In the last commit is changed the format of columns of the update, since I was testing. But it should not exist and I understand that it is not a solution to support postgresql.

Greetings.

**Comment by Pyzyryab** [comment]
I don't understand what's the problem with posgres. Can you post the errores, problems or related code?

By the way, I am probably not understanding correctly your problem, but you not be messing up with posgres code.

Whatever alternative you need in MySQL,  should be completly independent from postgresql code.

Are you maybe talking about the query macros? The approach is typically that postgre SQL is our base crate, so, whenever you need to do something with the queries, you must replace the problematic syntax in the executor, like we did in sqlserver.

This solution is far from perfect in terms of performance, but is the good one ATM  to keep the queries code completly decoupled from one driver vendor to another, si people can come and implement new drivers, like your case with MySQL.

By the way, we're open to upgrade our codebase, so we will appreciate any idea

**Comment by Pyzyryab** [comment]
> Good morning,
> I am starting to create test runs of select and update macros, without starting in querybuilder.
> I receive syntax errors on certain occasions, an example is in the names of the columns due to a quotation marks, in case of sql server does not affect, the upper or lower case, in case mysq_async does not allow me to keep them, and in case of postgresql if it can be the case and can be problematic, so I comment this problem to obtain a solution as effective as possible and not a horrible patch without having analyzed together.
> 
> In the last commit is changed the format of columns of the update, since I was testing. But it should not exist and I understand that it is not a solution to support postgresql.
> 
> Greetings.

@gbm25 by the way, you looked the upper-lower case standarization within Canyon for the queries. Can you help @0nSystem with its problem?

**Comment by gbm25** [comment]
> @gbm25 by the way, you looked the upper-lower case standarization within Canyon for the queries. Can you help @0nSystem with its problem?

If I remember correctly, by default we used the lowercase name for both table and column names, unless specified otherwise by the user.

Of course, to support both cases, the queries have to be prepared to search for those names taking into account that the names can contain uppercase letters because the table was created that way.

As I don't know what is the concrete problem of @0nSystem I will expose a small summary of the things related to table and column names in the different databases.

## Identifier Behavior in Different Databases

### PostgreSQL:
- Identifiers that are **not double-quoted** are folded to lowercase.
- Identifiers that **are double-quoted** retain their case and can be either reserved keywords or contain characters not generally allowed by the syntax rules.
- Examples:
<details>
  <summary>Click to see examples</summary>
  
  ### Table Creation:
```sql
CREATE TABLE "TEST" (
    "COLUMN1" VARCHAR(255),
    "COLUMN2" INT,
    "COLUMN3" DATE,
    "COLUMN4" BOOLEAN
);
```

### Sample Query:
```sql
INSERT INTO "TEST" ("COLUMN1", "COLUMN2", "COLUMN3", "COLUMN4") VALUES ('Sample', 123, '2023-08-14', TRUE);
SELECT * FROM "TEST" WHERE "COLUMN1" = 'Sample';
```
 
</details>

- [Reference documentation](https://www.postgresql.org/docs/current/sql-syntax-lexical.html#SQL-SYNTAX-IDENTIFIERS)

### Microsoft SQL Server:
- By default, SQL Server identifiers are **case-insensitive**.
- The `SET QUOTED_IDENTIFIER` option allows SQL Server to follow ISO rules regarding quotation mark delimiting identifiers and literal strings.
- When `SET QUOTED_IDENTIFIER` is `ON`, identifiers can be delimited by **double quotation marks** (`" "`) and can be either reserved keywords or contain characters not generally allowed by the syntax rules.
- Examples:
<details>
  <summary>Click to see examples</summary>
  
### Table Creation:
```sql
CREATE TABLE `TEST` (
    `COLUMN1` VARCHAR(255),
    `COLUMN2` INT,
    `COLUMN3` DATE,
    `COLUMN4` BOOLEAN
);
```

### Sample Query:
```sql
INSERT INTO `TEST` (`COLUMN1`, `COLUMN2`, `COLUMN3`, `COLUMN4`) VALUES ('Sample', 123, '2023-08-14', TRUE);
SELECT * FROM `TEST` WHERE `COLUMN1` = 'Sample';
```
 
</details>

- [Reference documentation](https://docs.microsoft.com/en-us/sql/t-sql/statements/set-quoted-identifier-transact-sql?view=sql-server-ver15)

### MySQL:
- Table names and column names are **case-insensitive on Windows** and **case-sensitive on Unix-based systems** by default.
- **Backticks** (` `` `) are used to quote identifiers, allowing for reserved words or special characters in names.
- Examples:
<details>
  <summary>Click to see examples</summary>
  
### Table Creation:
```sql
SET QUOTED_IDENTIFIER ON;
CREATE TABLE "TEST" (
    "COLUMN1" VARCHAR(255),
    "COLUMN2" INT,
    "COLUMN3" DATE,
    "COLUMN4" BIT
);
```

### Sample Query:
```sql
INSERT INTO "TEST" ("COLUMN1", "COLUMN2", "COLUMN3", "COLUMN4") VALUES ('Sample', 123, '2023-08-14', 1);
SELECT * FROM "TEST" WHERE "COLUMN1" = 'Sample';
```
 
</details>

- [Reference documentation](https://dev.mysql.com/doc/refman/8.0/en/identifiers.html)


Given the above, it is clear that we were able to use the same standard with PostgreSQL and Microsoft SQL Server, but it is not going to be possible for MySQL.
In the migrations this is not a problem because they are differentiated, in the rest of the code ... I'm not sure because I don't touch it so much.

It can also be that not all queries follow this format in Canyon-SQL, because when it was implemented in Microsoft SQL Server it gave a lot of problems and many times we did things a little bit blindly.

Anyway, if you have any doubt with the queries, let me know which one is failing or which part of the code is giving you problems and we will see it.

**Comment by 0nSystem** [comment]
> > @gbm25 by the way, you looked the upper-lower case standarization within Canyon for the queries. Can you help @0nSystem with its problem?
> 
> If I remember correctly, by default we used the lowercase name for both table and column names, unless specified otherwise by the user.
> 
> Of course, to support both cases, the queries have to be prepared to search for those names taking into account that the names can contain uppercase letters because the table was created that way.
> 
> As I don't know what is the concrete problem of @0nSystem I will expose a small summary of the things related to table and column names in the different databases.
> ## Identifier Behavior in Different Databases
> ### PostgreSQL:
> 
>     * Identifiers that are **not double-quoted** are folded to lowercase.
> 
>     * Identifiers that **are double-quoted** retain their case and can be either reserved keywords or contain characters not generally allowed by the syntax rules.
> 
>     * Examples:
> 
> 
> Click to see examples
> 
>     * [Reference documentation](https://www.postgresql.org/docs/current/sql-syntax-lexical.html#SQL-SYNTAX-IDENTIFIERS)
> 
> 
> ### Microsoft SQL Server:
> 
>     * By default, SQL Server identifiers are **case-insensitive**.
> 
>     * The `SET QUOTED_IDENTIFIER` option allows SQL Server to follow ISO rules regarding quotation mark delimiting identifiers and literal strings.
> 
>     * When `SET QUOTED_IDENTIFIER` is `ON`, identifiers can be delimited by **double quotation marks** (`" "`) and can be either reserved keywords or contain characters not generally allowed by the syntax rules.
> 
>     * Examples:
> 
> 
> Click to see examples
> 
>     * [Reference documentation](https://docs.microsoft.com/en-us/sql/t-sql/statements/set-quoted-identifier-transact-sql?view=sql-server-ver15)
> 
> 
> ### MySQL:
> 
>     * Table names and column names are **case-insensitive on Windows** and **case-sensitive on Unix-based systems** by default.
> 
>     * **Backticks** (`` `` ``) are used to quote identifiers, allowing for reserved words or special characters in names.
> 
>     * Examples:
> 
> 
> Click to see examples
> 
>     * [Reference documentation](https://dev.mysql.com/doc/refman/8.0/en/identifiers.html)
> 
> 
> Given the above, it is clear that we were able to use the same standard with PostgreSQL and Microsoft SQL Server, but it is not going to be possible for MySQL. In the migrations this is not a problem because they are differentiated, in the rest of the code ... I'm not sure because I don't touch it so much.
> 
> It can also be that not all queries follow this format in Canyon-SQL, because when it was implemented in Microsoft SQL Server it gave a lot of problems and many times we did things a little bit blindly.
> 
> Anyway, if you have any doubt with the queries, let me know which one is failing or which part of the code is giving you problems and we will see it.

Hello @gbm25  @Pyzyryab, 
With this you solve my problem, since I was referring to the name of standardization of the table and columns of postgres interferes with the implementation of mysql, but it is not a problem. I can solve it as a patch but I think we should look for a better way, than manipulating the query at the end prior to execution.

**Comment by damibazarra** [comment]
@0nSystem is there any progress on this? I have a personal project in mind with some friends and we would like to do some testing with this innovative implementation. Great job!

**Comment by 0nSystem** [comment]
> @0nSystem is there any progress on this? I have a personal project in mind with some friends and we would like to do some testing with this innovative implementation. Great job!

Hello @damibazarra , Thanks

In principle everything is almost ready, currently the implementation is functional and you can start using it, I have to fix MysqlOperator and clippy errors.

**Comment by 0nSystem** [comment]
The implementation is ready, if you see something that could be improved, do not hesitate to comment. @Pyzyryab @gbm25

**Comment by Pyzyryab** [comment]
> The implementation is ready, if you see something that could be improved, do not hesitate to comment. @Pyzyryab @gbm25

Hey @0nSystem I reviewed briefly your changes, and except some minor things, everything looks fine. Even more, you cleaned some parts of the cfg macros that I liked.

Please, consider to carefully read the comments left as notes, and implement the changes whenever you feel ready. And by the way, you can change the PR mode to open, since we are already preparing for the next release, that will include the `my_sql` implementation.

Thanks for your job :)

**Comment by 0nSystem** [comment]
Hey @Pyzyryab  later today, I'll check the comments, and let you know when I'm ready to merge.

**Comment by 0nSystem** [comment]
Hey @Pyzyryab later today, I'll check the comments, and let you know when I'm ready to merge.

**Review by Pyzyryab — APPROVED** [review]

**Pull request merged**